controllers/remote/cluster_cache.go

```diff
@@ -111,9 +113,45 @@ type Watcher interface {
 
 // watchInfo is used as a map key to uniquely identify a watch. Because predicates is a slice, it cannot be included.
 type watchInfo struct {
-	watcher      Watcher
-	kind         runtime.Object
-	eventHandler handler.EventHandler
+	watcher Watcher
+	gvk     schema.GroupVersionKind
+
+	// Comparing the eventHandler as an interface doesn't work because reflect.DeepEqual
+	// will assert functions are false if they are non-nil.
+	// Use a signature string representation instead as this can be compared.
+	// The signature function is expected to produce a unique output for each unique handler
+	// function that is passed to it.
+	// In combination with the watcher, this should be enough to identify unique watches.
+	eventHandlerSignature string
+}
+
+// eventHandlerSignature generates a unique identifier for the given eventHandler by
+// printing it to a string using "%#v".
+// Eg "&handler.EnqueueRequestsFromMapFunc{ToRequests:(handler.ToRequestsFunc)(0x271afb0)}"
+func eventHandlerSignature(h handler.EventHandler) string {
+	return fmt.Sprintf("%#v", h)
+}
+
+// gvkForObject finds the GroupVersionKind associated with the given object, if there is only a single such GVK.
+func (m *ClusterCacheTracker) gvkForObject(obj runtime.Object) (schema.GroupVersionKind, error) {
+	gvks, isUnversioned, err := m.scheme.ObjectKinds(obj)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+	if isUnversioned {
+		return schema.GroupVersionKind{}, fmt.Errorf("cannot watch unversioned type %T", obj)
+	}
+
+	if len(gvks) < 1 {
+		return schema.GroupVersionKind{}, fmt.Errorf("no group-version-kinds associated with type %T", obj)
+	}
+	if len(gvks) > 1 {
+		// this should only trigger for things like metav1.XYZ --
+		// normal versioned types should be fine
+		return schema.GroupVersionKind{}, fmt.Errorf(
+			"multiple group-version-kinds associated with type %T, refusing to guess at one", obj)
+	}
+	return gvks[0], nil
 }
```

Comment:
We can probably use https://github.com/kubernetes-sigs/controller-runtime/blob/c4893e0b992fb1e6e7da29acd692a0cdf9d7aec9/pkg/client/apiutil/apimachinery.go#L50-L69 instead

Reply:
Yeah we could, it's identical, except the error message "cannot create a new informer for the unversioned type %T", if that's acceptable then I'll put it to that and squash everything down

Reply:
That's probably fine, seems needs to be fixed upstream given that it's in the `apiutil` package. This function is also used in any of the controller runtime Client methods, so if it fails here it'd fail there as well